proxyd/server.go

```diff
@@ -175,7 +186,42 @@ func NewServer(
 		senderLim:              senderLim,
 		limExemptOrigins:       limExemptOrigins,
 		limExemptUserAgents:    limExemptUserAgents,
-	}, nil
+	}
+	s.updateLatestBlockNumber()
+	go s.pollLatestBlockNumber()
+	return s, nil
+}
+
+func (s *Server) pollLatestBlockNumber() {
+	ticker := time.NewTicker(2 * time.Second)
+	for {
+		select {
+		case <-ticker.C:
+			s.srvMu.Lock()
+			if s.shutdown {
+				s.srvMu.Unlock()
+				return
+			}
+			s.updateLatestBlockNumber()
+			s.srvMu.Unlock()
```

Comment:
It looks like we're holding this mutex across the entire function. It might make more sense to split the `updateLatestBlockNumber` function into a `fetchLatestBlockNumber() (uint64, error)` and then a `updateLatestBlockNumber(uint64)` that contains the locking logic. Something like:

```golang
func (s *Server) pollLatestBlockNumber() {
	ticker := time.NewTicker(2 * time.Second)
	for {
		select {
		case <-ticker.C:
			bn, err := s.fetchLatestBlockNumber()
            if err != nil {
                s.updateLatestBlockNumber(bn)
            }
		}
	}
}

func (s *Server) updateLatestBlockNumber(bn uint64) {
    s.srvMu.Lock()
    defer s.srvMu.Unlock()
	s.latestBlockNumber.Store(bn)
}

func (s *Server) fetchLatestBlockNumber() (uint64, error) {
	req := json.RawMessage("{\"id\":0,\"jsonrpc\":\"2.0\",\"method\":\"eth_blockNumber\"}")
	res, _, err := s.handleBatchRPC(context.Background(), []json.RawMessage{req}, func(method string) bool { return false }, false)
	if err != nil {
		log.Error("error requesting latest block number", "err", err)
		return uint64(0), err
	}
	bns := res[0].Result.(string)
	bn, err := hexutil.DecodeUint64(bns)
	if err != nil {
		log.Error("error decoding hex block number", "err", err)
		return uint64(0), err
	}
	return bn, nil
}
```

Reply:
SGTM, will dismiss this, still works the current way so no concern here :)